### PR TITLE
CRF layer v3.0

### PR DESCRIPTION
**Pull request by gabrieldemarmiesse**
With a subclassing approch, we have a nicer API and it's very flexible.

Works only with TF 2.2+

@howl-anderson  for the review and the CLA

The plan is to show users how to do the subclassing for the CRF. We shouldn't provide and API to save them some code there because it's going to become very complex to design a good API and to maintain it later on.

So the CRF layer is a public API and for the CRF loss, we give a good tutorial about subclassing.

Quick tutorial right now:

```python
import tensorflow as tf
from tensorflow_addons.layers.crf import CRF
from tensorflow_addons.text.crf import crf_log_likelihood

def unpack_data(data):
    if len(data) == 2:
        return data[0], data[1], None
    elif len(data) == 3:
        return data
    else:
        raise TypeError("Expected data to be a tuple of size 2 or 3.")


class ModelWithCRFLoss(tf.keras.Model):
    """Wrapper around the base model for custom training logic."""

    def __init__(self, base_model):
        super().__init__()
        self.base_model = base_model

    def call(self, inputs):
        return self.base_model(inputs)

    def compute_loss(self, x, y, sample_weights, training=False):
        y_pred = self(x, training=training)
        _, potentials, sequence_length, chain_kernel = y_pred

        crf_loss = -crf_log_likelihood(potentials, y, sequence_length, chain_kernel)[0]

        if sample_weights is not None:
            crf_loss = crf_loss * sample_weights

        return tf.reduce_mean(crf_loss), sum(self.losses)

    def train_step(self, data):
        x, y, sample_weight = unpack_data(data)

        with tf.GradientTape() as tape:
            crf_loss, internal_losses = self.compute_loss(
                x, y, sample_weight, training=True
            )
            total_loss = crf_loss + internal_losses

        gradients = tape.gradient(total_loss, self.trainable_variables)
        self.optimizer.apply_gradients(zip(gradients, self.trainable_variables))

        return {"crf_loss": crf_loss, "internal_losses": internal_losses}

    def test_step(self, data):
        x, y, sample_weight = unpack_data(data)
        crf_loss, internal_losses = self.compute_loss(x, y, sample_weight)
        return {"crf_loss_val": crf_loss, "internal_losses_val": internal_losses}


x_np, y_np = get_test_data()

x_input = tf.keras.layers.Input(shape=x_np.shape[1:])
crf_outputs = CRF(5)(x_input)
base_model = tf.keras.Model(x_input, crf_outputs)
model = ModelWithCRFLoss(base_model)

model .compile("adam")
model .fit(x=x_np, y=y_np)
model .evaluate(x_np, y_np)
model .predict(x_np)
model.save("my_model.tf")
```

If some users want to try this feature before it's merged, we have [some wheels available](https://github.com/tensorflow/addons/actions/runs/88454622)

### CODE REVIEW

File: tensorflow_addons/layers/tests/crf_test.py [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+        # we now add the CRF loss:
+        crf_loss = -crf_log_likelihood(potentials, y, sequence_length, chain_kernel)[0]
+
+        if sample_weights is not None:
+            crf_loss = crf_loss * sample_weights
+
+        return tf.reduce_mean(crf_loss), sum(self.losses)
+
+    def train_step(self, data):
+        x, y, sample_weight = unpack_data(data)
```
Comment (elazarg) [inline tensorflow_addons/layers/tests/crf_test.py:146]: Inconsistent naming - it's `sample_weights`, plural, in `compute_loss()`

File: tensorflow_addons/layers/tests/crf_test.py [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+
+        gradients = tape.gradient(total_loss, self.trainable_variables)
+        self.optimizer.apply_gradients(zip(gradients, self.trainable_variables))
+
+        return {"crf_loss": crf_loss, "internal_losses": internal_losses}
+
+    def test_step(self, data):
+        x, y, sample_weight = unpack_data(data)
+        crf_loss, internal_losses = self.compute_loss(x, y, sample_weight)
+        return {"crf_loss_val": crf_loss, "internal_losses_val": internal_losses}
```
Comment (elazarg) [inline tensorflow_addons/layers/tests/crf_test.py:162]: prefix `val_` is [already added](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/training.py#L930)

**Comment by YanZhu1105** [comment]
Hi,

I am relatively new to Tensorflow. Just wondering I can't seem to find tensorflow_addons.layers.crf?
And when I 'from tensorflow_addons.layers.crf im port CRF', the error shows no module named 'tensorflow_addons.layers.crf.

Thanks

**Comment by YanZhu1105** [comment]
Hi,

Is it possible to add some feature to allow sample weights?

Thanks!

**Comment by YanZhu1105** [comment]
Hi, 
Is it possible to include an example with sample weights?
Also an example where the input of model.fit is a generator which yield (x, y, sample_weight) for eatch batch?

Thanks!

File: tensorflow_addons/layers/crf.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+            raw_input_shape = tf.slice(input_energy_shape, [0], [2])
+            alt_mask = tf.ones(raw_input_shape)
+
+            sequence_length = self.mask_to_sequence_length(alt_mask)
+
+        return sequence_length
+
+    def mask_to_sequence_length(self, mask):
+        """compute sequence length from mask."""
+        sequence_length = tf.cast(tf.reduce_sum(tf.cast(mask, tf.int8), 1), tf.int64)
```
Comment (ndrewl) [inline tensorflow_addons/layers/crf.py:142]: Here sums are computed using `tf.int8` which will overflow on sequences longer than 127, which will produce negative sequence lengths. Maybe we cast the mask to `tf.int64` right away, and then the outer cast will be unnecessary?

**Comment by luozhouyang** [comment]
Any updates?

**Comment by luozhouyang** [comment]
Any updates?

**Comment by ndrewl** [comment]
@gabrieldemarmiesse, are you still going to work on this PR?

**Comment by gabrieldemarmiesse** [comment]
Sorry, I'm very busy nowadays, somebody else is more than welcome to take this branch and open a new pull request with it :)

**Comment by FakeCola** [comment]
@gabrieldemarmiesse Hey! I'm working on a model with CRF layer recently and your solution here is very helpful. I'd gladly help finish this PR if you are not available currently ;)

**Comment by gabrieldemarmiesse** [comment]
I'm glad it helped you! Feel free to pull this branch into your fork and open a new PR :)

**Comment by gabrieldemarmiesse** [comment]
Closing in favor of #1999

**Comment by DachuanZhao** [comment]
> ```python
> base_model
> ```

Hi ~ How to use this code to build a bi-lstm-crf model ? like this ? 
```
x_input = tf.keras.layers.Input(shape=x_np.shape[1:])
bilstm_output = bi_lstm_model(x_input)
crf_outputs = CRF(5)(bilstm_output )
base_model = tf.keras.Model(x_input, crf_outputs)
model = ModelWithCRFLoss(base_model)
```

**Comment by DachuanZhao** [comment]
> With a subclassing approch, we have a nicer API and it's very flexible.
> 
> Works only with TF 2.2+
> 
> @howl-anderson for the review and the CLA
> 
> The plan is to show users how to do the subclassing for the CRF. We shouldn't provide and API to save them some code there because it's going to become very complex to design a good API and to maintain it later on.
> 
> So the CRF layer is a public API and for the CRF loss, we give a good tutorial about subclassing.
> 
> Quick tutorial right now:
> 
> ```python
> import tensorflow as tf
> from tensorflow_addons.layers.crf import CRF
> from tensorflow_addons.text.crf import crf_log_likelihood
> 
> def unpack_data(data):
>     if len(data) == 2:
>         return data[0], data[1], None
>     elif len(data) == 3:
>         return data
>     else:
>         raise TypeError("Expected data to be a tuple of size 2 or 3.")
> 
> 
> class ModelWithCRFLoss(tf.keras.Model):
>     """Wrapper around the base model for custom training logic."""
> 
>     def __init__(self, base_model):
>         super().__init__()
>         self.base_model = base_model
> 
>     def call(self, inputs):
>         return self.base_model(inputs)
> 
>     def compute_loss(self, x, y, sample_weights, training=False):
>         y_pred = self(x, training=training)
>         _, potentials, sequence_length, chain_kernel = y_pred
> 
>         crf_loss = -crf_log_likelihood(potentials, y, sequence_length, chain_kernel)[0]
> 
>         if sample_weights is not None:
>             crf_loss = crf_loss * sample_weights
> 
>         return tf.reduce_mean(crf_loss), sum(self.losses)
> 
>     def train_step(self, data):
>         x, y, sample_weight = unpack_data(data)
> 
>         with tf.GradientTape() as tape:
>             crf_loss, internal_losses = self.compute_loss(
>                 x, y, sample_weight, training=True
>             )
>             total_loss = crf_loss + internal_losses
> 
>         gradients = tape.gradient(total_loss, self.trainable_variables)
>         self.optimizer.apply_gradients(zip(gradients, self.trainable_variables))
> 
>         return {"crf_loss": crf_loss, "internal_losses": internal_losses}
> 
>     def test_step(self, data):
>         x, y, sample_weight = unpack_data(data)
>         crf_loss, internal_losses = self.compute_loss(x, y, sample_weight)
>         return {"crf_loss_val": crf_loss, "internal_losses_val": internal_losses}
> 
> 
> x_np, y_np = get_test_data()
> 
> x_input = tf.keras.layers.Input(shape=x_np.shape[1:])
> crf_outputs = CRF(5)(x_input)
> base_model = tf.keras.Model(x_input, crf_outputs)
> model = ModelWithCRFLoss(base_model)
> 
> model .compile("adam")
> model .fit(x=x_np, y=y_np)
> model .evaluate(x_np, y_np)
> model .predict(x_np)
> model.save("my_model.tf")
> ```
> 
> If some users want to try this feature before it's merged, we have [some wheels available](https://github.com/tensorflow/addons/actions/runs/88454622)

Hi ~ How to use this code to build a bi-lstm-crf model ? like this ? 
```
x_input = tf.keras.layers.Input(shape=x_np.shape[1:])
bilstm_output = bi_lstm_model(x_input)
crf_outputs = CRF(5)(bilstm_output)
base_model = tf.keras.Model(x_input, crf_outputs)
model = ModelWithCRFLoss(base_model)
```

**Comment by xuxingya** [comment]
I build a crf tool for TF2, you can refer to [tf2crf](https://github.com/xuxingya/tf2crf). And pip install tf2crf.

**Comment by DachuanZhao** [comment]
> I build a crf tool for TF2, you can refer to [tf2crf](https://github.com/xuxingya/tf2crf). And pip install tf2crf.

What's the difference between your CRF layer and ```tensorflow_addons.layers.crf``` ？

**Comment by xuxingya** [comment]
Because during a longtime the tfa.layers.crf has problems with mixed
precision and other bugs. So I implemented a CRF layer using the CRF
functions in tfa.text.crf.
